### PR TITLE
Add new endpoint API ComposeTree

### PR DESCRIPTION
**Pull request by simozhan**
JIRA: PDC-1100

### CODE REVIEW

**Comment by xychu** [comment]
LGTM

File: pdc/apps/compose/views.py [hunk 1]
```diff
@@ -1399,3 +1402,142 @@ def list(self, request):
...
+    """
+    API endpoint that allows querying compose-variant-arch relevant to location.
+
+    ##Test tools##
+
+    You can use ``curl`` in terminal, with -X _method_ (GET|POST|PUT|DELETE),
+    -d _data_ (a json string). or GUI plugins for
+    browsers, such as ``RESTClient``, ``RESTConsole``.
+    """
+    queryset = ComposeTree.objects.all()
```
Comment (ycheng-aa) [inline pdc/apps/compose/views.py]: If some other related objects also need to be accessed, maybe could use select_related() to fetch them with on DB hit.

File: pdc/apps/compose/views.py [hunk 2]
```diff
@@ -1399,3 +1402,142 @@ def list(self, request):
...
+
+        updatable_keys = set(['scheme', 'url', 'synced_content'])
+        if set(request.data.keys()) - updatable_keys:
+            return Response(status=status.HTTP_400_BAD_REQUEST,
+                            data={'detail': 'Only these properties can be updated: %s'
+                                  % ', '.join(updatable_keys)})
+
+        return super(ComposeTreeViewSet, self).update(request, *args, **kwargs)
+
+    def bulk_update(self, *args, **kwargs):
```
Comment (ycheng-aa) [inline pdc/apps/compose/views.py]: Is it necessary to implement it? I think with BulkRouter, the default implementation is like this.

**Comment by ycheng-aa** [comment]
Generally looks good to me.

File: pdc/apps/compose/models.py [hunk 3]
```diff
@@ -505,3 +505,41 @@ class Meta:
...
+
+
+class Scheme(models.Model):
+    name = models.CharField(max_length=50, unique=True)
+
+
+class ComposeTree(models.Model):
+    compose             = models.ForeignKey("Compose")
+    variant             = models.ForeignKey("Variant")
+    arch                = models.CharField(max_length=50)
```
Comment (sochotnicky) [inline pdc/apps/compose/models.py]: I think these should be one of currently defined arches and not just any string.

File: pdc/apps/compose/models.py [hunk 4]
```diff
@@ -505,3 +505,41 @@ class Meta:
...
+    location            = models.ForeignKey("Location")
+    scheme              = models.ForeignKey("Scheme")
+    url                 = models.CharField(max_length=100)
```
Comment (sochotnicky) [inline pdc/apps/compose/models.py]: max length 100 is definitely too restrictive. Why should the url be that short? We would certainly hit this limit

File: pdc/apps/compose/models.py (hunk 3 above)
Comment (simozhan) [inline pdc/apps/compose/models.py]: Ok. fixed.

File: pdc/apps/compose/models.py (hunk 4 above)
Comment (simozhan) [inline pdc/apps/compose/models.py]: Yes. it is too restrictive. change it to 255.

File: pdc/apps/compose/views.py (hunk 2 above)
Comment (simozhan) [inline pdc/apps/compose/views.py]: Yes, the default implementation is like this. but the doc is not good enough, this api only support partial update.

**Comment by lubomir** [comment]
Looks fine to me.

**Pull request closed without merging**